perf(sql):  JIT-optimise `IN(numeric values)` operator

zebs97 opened the pull request:
Closes #4659

## Proposed solution

The idea for my implementation is stop the traversing algorithm at the root node of the predicate (the node representing the IN operator), and expand the IN operator as described in the issue in when visiting that node.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/questdb/questdb?pullRequest=4661) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/questdb/questdb?pullRequest=4661) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/questdb/questdb?pullRequest=4661) it.</sub>

puzpuzpuz commented:
A few tests seem to fail, most likely due to changed exec plans. Could you check?
```
Tests in error: 
  ClickBenchTest.testExecutionPlans:693->AbstractCairoTest.assertMemoryLeak:954->AbstractCairoTest.assertMemoryLeak:971->AbstractCairoTest.lambda$assertMemoryLeak$8:974->lambda$testExecutionPlans$0:698->AbstractCairoTest.assertPlanNoLeakCheck:1607->AbstractCairoTest.getPlanFactory:1933->AbstractCairoTest.select:1541->AbstractCairoTest.select:1537 » NoSuchElement
  TimestampQueryTest.testTimestampInDay1orDay2:781->AbstractCairoTest.assertQuery:1072->AbstractCairoTest.assertMemoryLeak:954->AbstractCairoTest.assertMemoryLeak:971->AbstractCairoTest.lambda$assertMemoryLeak$8:974->AbstractCairoTest.lambda$assertQuery$10:1072->AbstractCairoTest.assertQueryNoLeakCheck:1130->AbstractCairoTest.printSqlResult:1440->AbstractCairoTest.select:1541->AbstractCairoTest.select:1537 » NoSuchElement
```

puzpuzpuz commented on core/src/test/java/io/questdb/test/jit/CompiledFilterIRSerializerTest.java:
```diff
@@ -784,6 +784,12 @@ public void testVarcharNullConstant() throws Exception {
     }
 
+    @Test
+    public void testIn() throws Exception {
```
Would be great to have some variations in the test, different number of input values, more complex filters, separate tests for error cases like `anint in ()`.

zebs97 commented:
@puzpuzpuz I fixed the failing tests and added new test cases.

After some debugging, I found out the cause of the failure was an error in the implementation. When serializing the IN node, I was only calling `visit` on the arg nodes. This was insufficient to correctly handle negated constants, as these are serialized as stubs in the `descend` method and then backfilled after leaving the predicate.

After some consideration, I think the best solution here is to actually traverse the nodes when serializing the IN clause. Simply calling both `descend` and `visit` (as I implemented in a previous commit) is enough for negated constants, but not for non-constant negated values, which require full node traversal. For that purpose, I create a new `PostOrderTreeTraversalAlgo` to have a fresh stack for each node within the IN serialization function.

Please let me know if you can think of any other case worth trying which might fail.

puzpuzpuz commented:
> Please let me know if you can think of any other case worth trying which might fail.

We should test `asymbol in (select asymbol from tab limit 1)` operators - serialization should fail for them. Also, would be great to test filters with bind variables like `along in ($1, $2, $3)` (see `testBindVariablesMixed` as an example).

puzpuzpuz reviewed (dismissed):
Many thanks for the contribution!

zebs97 commented:
Thank you for your time reviewing it 😄

bluestreak01 commented:
This PR regresses existing behaviour. There are test failures:

![image](https://github.com/questdb/questdb/assets/7276403/572c7253-6060-43f6-b7c2-545601171dec)

puzpuzpuz commented:
@zebs97 could you check the failed tests?

zebs97 commented:
@puzpuzpuz @bluestreak01 Sorry, I was OOO this week and couldn't work on this. I'm preparing a PR fixing all the tests.

zebs97 commented:
@puzpuzpuz apart from fixing the exec plan tests that were failing, my last commits include the following changes:

- Support IN serialization when IN node is not the top-level predicate node. This happens when we have a NOT operator e.g. `anint NOT IN (1, 2, 3)`. This bug was causing the `StackOverflowError` in the failing tests.
- Fix the resulting IR for IN operators. In my initial implementation the resulting IR appended an `(||)` operator after every two `(=)` comparisons. I checked and this is not how multiple OR are handled by the IR serializer, as the following test passes:
```
@Test
public void test() throws Exception {
    serialize("anint = 1 OR anint = 2 OR anint = 3");
    assertIR("(i32 3L)(i32 anint)(=)(i32 2L)(i32 anint)(=)(i32 1L)(i32 anint)(=)(||)(||)(ret)");
}
```
I changed all IN tests accordingly.

zebs97 commented:
There's a seemingly unrelated test failing `testJsonQueryErrorOnDataUnavailableEventNeverFired`. When I execute it locally it seems to be flaky, sometimes it passes and sometimes it fails.

bluestreak01 commented:
hi @zebs97, this is ok:

>There's a seemingly unrelated test failing testJsonQueryErrorOnDataUnavailableEventNeverFired. When I execute it locally it seems to be flaky, sometimes it passes and sometimes it fails.

test fails due to SQL timeout, which does happen occasionally on CI (it is very slow).

puzpuzpuz commented:
> * I checked and this is not how multiple OR are handled by the IR serializer, as the following test passes

Considering this, could you add a test that involves a filter with IN and verifies that all IN values are in the result set to, say, `ParallelFilterTest`? Also, it would be great to add a test similar to `CompiledFilterIRSerializerTest#testInSubSelect()` that would verify that the compiler doesn't support like this one: `atimestamp IN '2020-01-01'`

zebs97 commented:
Hi @puzpuzpuz, about your second point: 

> Also, it would be great to add a test similar to CompiledFilterIRSerializerTest#testInSubSelect() that would verify that the compiler doesn't support like this one: atimestamp IN '2020-01-01'

I checked and currently the compilation fails because the compiler doesn't support string literals in the context of a numerical predicate type. Is this enough for now?

puzpuzpuz commented:
> I checked and currently the compilation fails because the compiler doesn't support string literals in the context of a numerical predicate type. Is this enough for now?

It would be great to add a test for that.

zebs97 commented:
Added the requested tests:
- Compilation fails with `atimestamp IN <string_literal>` (`CompiledFilterIRSerializerTest#testTimestampIn`)
- IN query resolves appropriate records both with and without JIT (`ParallelFilterTest#testInJitEnabled` and `ParallelFilterTest#testInJitDisabled`)

zebs97 commented:
@puzpuzpuz checking the test failures. 

#4658 changed the beahaviour around `atimestamp IN <timestamp_literal>'. This optimization now requires a more explicit distinction between the two IN operators. I checked the resulting `ExpressionNode`s for the two operators, and since we can't use the `token` to differentiate them, I think we need to rely on the `type` attribute.

- `atimestamp IN <timestamp_literal>` has a node with `token` = `"in"` and `type` = `32` (SET_OPERATION)
- `atimestamp IN (<timestamp_literal>[, <timestamp_literal>]*)` has a node with `token` = `"in"` and `type` = `8` (FUNCTION)

So for the operation serialization method, we can include the condition `type == ExpressionNode.FUNCTION` so that the SET_OPERATION node would result in an invalid operator. 

I checked this locally and it seems to work, I'm executing the full tests suite before pushing the changes to make sure nothing else is broken.

BTW can you share the semantics behind the ExpressionNode types? I would've expected `atimestamp IN (...)` to be a SET_OPERATION.

puzpuzpuz commented:
For now, it's fine not to support both `atimestamp IN <timestamp_literal>` and `atimestamp IN (<timestamp_literals>)`. The former one is a more compact way to write `atimestamp BETWEEN <timestamp_literal1> AND <timestamp_literal2>' (it's more flexible than that since you can also [specify ranges](https://questdb.io/docs/reference/operators/date-time/#in-timerangewithmodifier)). The latter is nothing more than `atimestamp = <timestamp_literal1> OR ... OR atimestamp = <timestamp_literalN>`.

> BTW can you share the semantics behind the ExpressionNode types? I would've expected `atimestamp IN (...)` to be a SET_OPERATION.

AFAIR that's nothing more but an internal convention. `atimestamp IN (...)` potentially has multiple arguments, so it's implicitly considered to be a vararg function.

puzpuzpuz reviewed (approved):
Thanks for the updates!

puzpuzpuz commented:
@zebs97 many thanks for your contribution!
